chain/ethereum/client.go

```diff
@@ -0,0 +1,136 @@
+package ethereum
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/renproject/multichain/api/account"
+	"github.com/renproject/multichain/api/address"
+	"github.com/renproject/multichain/api/contract"
+	"github.com/renproject/pack"
+)
+
+const (
+	// DefaultClientRPCURL is the RPC URL used by default, to interact with the
+	// ethereum node.
+	DefaultClientRPCURL = "http://127.0.0.1:8545/"
+)
+
+// Client holds rpc URL to connect to a ethereum node, and the underlying
+// RPC client instance.
+type Client struct {
+	rpcURL    string
+	ethClient *ethclient.Client
+}
+
+// NewClient creates and returns a new JSON-RPC client to the Ethereum node
+func NewClient(rpcURL string) (*Client, error) {
+	client, err := ethclient.Dial(rpcURL)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		rpcURL, client,
+	}, nil
+}
+
+// LatestBlock returns the block number at the current chain head.
+func (client *Client) LatestBlock(ctx context.Context) (pack.U64, error) {
+	header, err := client.ethClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return pack.NewU64(0), err
+	}
+	return pack.NewU64(header.Number.Uint64()), nil
+}
+
+// Tx returns the transaction uniquely identified by the given transaction
+// hash. It also returns the number of confirmations for the transaction.
+func (client *Client) Tx(ctx context.Context, txID pack.Bytes) (account.Tx, pack.U64, error) {
+	tx, pending, err := client.ethClient.TransactionByHash(ctx, common.BytesToHash(txID))
+	if err != nil {
+		return nil, pack.NewU64(0), err
+	}
+	chainID, err := client.ethClient.ChainID(ctx)
+	if err != nil {
+		return nil, pack.NewU64(0), err
+	}
+	header, err := client.ethClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return nil, pack.NewU64(0), err
+	}
+	newTx := Tx{
+		tx,
+		types.MakeSigner(&params.ChainConfig{
+			ChainID: chainID,
+		}, header.Number),
+	}
+	if pending {
+		return &newTx, 0, nil
+	}
+	receipt, err := client.ethClient.TransactionReceipt(ctx, common.BytesToHash(txID))
+	if err != nil {
+		return nil, pack.NewU64(0), err
+	}
+	if receipt == nil {
+		return &newTx, 0, nil
+	}
+	return &newTx, pack.NewU64(header.Number.Uint64() - receipt.BlockNumber.Uint64()), nil
```

Comment:
The problem I see here is that, the transaction could have been a legacy tx, i.e. included even before the EIP-155 triggering block number. Using the chain head as the block number while instantiating the `signer` may result in issues with respect to calculating the `From()` for this tx.

Also, since every RPC call takes time to be resolved, we should make RPC calls in an order which allows us to return at the earliest. For instance, we don't need the chain head unless we know the transaction has been mined (we're using the header only to calculate the block confirmations for the tx). So making the header call before sorting out the other constraints adds more time for the function to return.

We could try something like:
```suggestion
// Tx returns the transaction uniquely identified by the given transaction hash.
// It also returns the number of confirmations for the transaction.
func (client *Client) Tx(ctx context.Context, txID pack.Bytes) (account.Tx, pack.U64, error) {
	// Get the transaction.
	tx, pending, err := client.ethClient.TransactionByHash(ctx, common.BytesToHash(txID))
	if err != nil {
		return nil, pack.NewU64(0), err
	}
	chainID, err := client.ethClient.ChainID(ctx)
	if err != nil {
		return nil, pack.NewU64(0), err
	}
	// If the transaction is still pending, we can use the EIP-155 signer.
	pendingTx := Tx{
		ethTx: tx,
		signer: types.NewEIP155Signer(chainID),
	}
	if pending {
		return &pendingTx, 0, nil
	}

	// Get the transaction receipt. Receipts are not available for pending tx.
	receipt, err := client.ethClient.TransactionReceipt(ctx, common.BytesToHash(txID))
	if err != nil {
		return nil, pack.NewU64(0), err
	}
	if receipt == nil {
		return &pendingTx, 0, nil
	}

	// Here we know that the tx had been included in a block. Instantiate the
	// signer appropriately using the tx receipt's block number.
	confirmedTx := Tx{
		ethTx: tx,
		signer: types.MakeSigner(&params.ChainConfig{ ChainID: chainID }, receipt.BlockNumber),
	}
	header, err := client.ethClient.HeaderByNumber(ctx, nil)
	if err != nil {
		return nil, pack.NewU64(0), err
	}
	confs := pack.NewU64(header.Number.Uint64() - receipt.BlockNumber.Uint64())

	return &confirmedTx, confs, nil
}
```